odl/discr/discr_ops.py

```diff
@@ -130,110 +144,212 @@ def finite_diff(f, out=None, axis=0, dx=1.0, edge_order=2,
 
     method, method_in = str(method).lower(), method
     if method not in ('central', 'forward', 'backward'):
-        raise ValueError("method '{}' has to be central, forward or backward"
+        raise ValueError('method {} is not central, forward or backward'
                          ''.format(method_in))
 
+    if isinstance(padding, (int, float)) or padding in (None, 'edge'):
+        if isinstance(padding, int):
```

Comment:
This will fail for any non-core data type (e.g. `np.int32` on Linux 64-bit):
```python
>>> x = np.ones(1, dtype='int32')
>>> isinstance(x[0], int)
False
```
Also don't try to handle all cases at once. It's better to start with `None`, then the first real case float, work through that, then each string choice, and finally a `ValueError` in the `else` clause. For the first case you can use a `try ... except` fence, i.e.
```python
try:
    padding = float(padding)
except TypeError:
    pass
else:
    # Do your thing with the float value
```